src/nvim/insexpand.c

```diff
@@ -2528,6 +2533,15 @@ static void ins_compl_add_dict(dict_T *dict)
   }
 }
 
+/// Save extmarks in "compl_orig_text" so that they may be restored when the
+/// completion is cancelled, or the original text is completed.
+static void save_orig_extmarks(void)
+{
+  compl_orig_extmarks = extmark_get(curbuf, 0, curwin->w_cursor.lnum - 1, compl_col,
```

Comment:
this copies and restores both ends of a paired extmark which is not what we want (this is going to make a bigger difference later, with proper extranges).

Instead I'd suggest refactoring `u_extmark_copy` into two parts: a smaller part which just pushes `ExtmarkUndoObject` into a `extmark_undo_vec_t *` taken as a pointer, and a wrapper which does the ` u_force_get_undo_header` part. (or perhaps, just move that to the call site, as there only is one)

Then we can make sure that the same underlying rules for saving/storing extmarks is the same as undo, even though we are not using undo headers here.

Reply:
Done (I think).